### PR TITLE
Add a generic shape primitive

### PR DESCRIPTION
**Pull request by jherico**
* provides a simpler means of adding primitives, since each new primitive will not require a new entity
 * Presumably we will still want to bump the entity version each time we add support for a new shape, even though technically we won't need to
* Fixes changes the 'sphere' shape used in physics calculations to ellipsoid, so that elongated spheres no longer have collision behavior as if they were giant spheres.
* Currently supports icosahedron, tetrahedron, octahedron and dodecahedron in addition to sphere and cube.

Issues:

* Unsupported types will parse, but will be rendered as spheres.
* Performance should be improved with instanced rendering.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7aedabe9bd55c8d018c6c50ac526bce015956bd3 available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 00bf5c0298711fada9b7b69bfba3726e04e801d9 available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f397c9b6c364ebfc8d245a56a34f2474b3954455 available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4f93a8664ee1b2297ae9016854e4027bf6a2f0d5 available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 28fc71caad72d263054738b3d71f74f1e20ab4f9 available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

**Comment by jherico** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 28fc71caad72d263054738b3d71f74f1e20ab4f9 available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

**Comment by kevinmthomas-carpool** [comment]
testing

**Comment by kevinmthomas-carpool** [comment]
## QA Test
**QA Tester**: KevinMThomas
**localhost 'createTestBlocks.js' Result**: **localhost 'createTestBlocks.js'** SUCCESS
**localhost ATP Upload + Edit Result**: **localhost ATP Upload + Edit** SUCCESS
**Echo Server Audio**: **Echo Server Audio** SUCCESS
**localhost 'harmonicOscillator.js' Result**: **localhost 'harmonicOscillator.js'** FAILURE
**localhost 'reverbTest.js' Result**: **localhost 'reverbTest.js'** SUCCESS
**localhost 'playaPerformanceTest.js' Result**: **localhost 'playaPerformanceTest.js'** SUCCESS
**playa 'playaPerformanceTest.js' Result**: **playa 'playaPerformanceTest.js'** FAILURE
**cellscience 'playaPerformanceTest.js' Result**: **cellscience 'playaPerformanceTest.js'** FAILURE
**QA Pass**: **QA Pass** FAILURE
**Additional Test Results + Observations**: Failure. Dramatically slow Render and Present New Rates, ```playa``` and ```cellscience``` domains refused to load and said incompatible server versions.  Successful view of avatars as I was able to speak and see Caitlyn at ```playa``` however no content loaded.  Stats also showed data however nothing loaded.

**Domain** |localhost |Playa |cellscience
---|---|---|---
**Render Rate**: |23 |15 |22
**Present Rate**: |60 |60 |60
**Present New Rate**: |24 |18 |22
**CPU Buffer Count**: |2936 |273 |211
**GPU Buffer Count**: |372 |135 |109
**CPU Memory Usage**: |29355200 |8432660 |6338684
**GPU Memory Usage**: |16005384 |8544984 |6631584
**CPU Texture Count**: |167 |70 |50
**GPU Texture Count**: |77 |61 |41
**GPU Transfer Count**: |0 |0 |0
**CPU Texture Memory Usage**: |210967876 |15022388 |15022388
**GPU Texture Memory Usage**: |327200171 |402814034 |160398590
**GPU Virtual Memory Usage**: |327200171|402814034|160398590
**Triangle Count**: |444558 |67479 |67479
**Triangle Rate**: |148186000 |16869750 |22493000
**API Drawcall Count**: |122 |14 |14
**Drawcall Count**: |117 |14 |14
**Drawcall Rate**: |39000 |3500 |4666
**State Changes Texture Count**:|104 |25 |24
**Pipeline Count**: |27 |14 |14
**Item Opaques**: |133 |7 |7
**Item Translucents**: |11 |2 |1
**Item Lights**: |24 |0 |0
**Timing Opaques**: |1153 |145 |147
**Timing Translucents**: |57 |10 |7
**Timing Lighting**: |66 |27 |29
**Render Frame**: |1865 |504 |562

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp [hunk 1]
```diff
@@ -39,6 +39,7 @@
 #include "RenderableLineEntityItem.h"
 #include "RenderablePolyVoxEntityItem.h"
 #include "RenderablePolyLineEntityItem.h"
+#include "RenderableShapeEntityItem.h"
```
Comment (zzmp) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: Should this remove the RenderableBox/RenderableSphere?

File: libraries/entities-renderer/src/RenderableShapeEntityItem.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+#include <render-utils/simple_vert.h>
+#include <render-utils/simple_frag.h>
+
+// Sphere entities should fit inside a cube entity of the same size, so a sphere that has dimensions 1x1x1 
+// is a half unit sphere.  However, the geometry cache renders a UNIT sphere, so we need to scale down.
+static const float SPHERE_ENTITY_SCALE = 0.5f;
+
+static GeometryCache::Shape MAPPING[entity::NUM_SHAPES] = {
+    GeometryCache::Triangle, // Triangle
```
Comment (zzmp) [inline libraries/entities-renderer/src/RenderableShapeEntityItem.cpp]: You don't need the comments next to these, they are self-commenting.

File: libraries/entities-renderer/src/RenderableShapeEntityItem.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    Q_ASSERT(getType() == EntityTypes::Shape);
+    Q_ASSERT(args->_batch);
+
+    if (!_procedural) {
+        _procedural.reset(new Procedural(getUserData()));
+        _procedural->_vertexSource = simple_vert;
+        _procedural->_fragmentSource = simple_frag;
+        _procedural->_state->setCullMode(gpu::State::CULL_NONE);
+        _procedural->_state->setDepthTest(true, true, gpu::LESS_EQUAL);
+        _procedural->_state->setBlendFunction(false,
```
Comment (zzmp) [inline libraries/entities-renderer/src/RenderableShapeEntityItem.cpp]: Shouldn't this be based on the transparency of the RenderableShapeEntityItem?

File: libraries/entities-renderer/src/RenderableShapeEntityItem.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            gpu::State::SRC_ALPHA, gpu::State::BLEND_OP_ADD, gpu::State::INV_SRC_ALPHA,
+            gpu::State::FACTOR_ALPHA, gpu::State::BLEND_OP_ADD, gpu::State::ONE);
+    }
+
+    gpu::Batch& batch = *args->_batch;
+    glm::vec4 sphereColor(toGlm(getXColor()), getLocalRenderAlpha());
```
Comment (zzmp) [inline libraries/entities-renderer/src/RenderableShapeEntityItem.cpp]: s/sphereColor/color

File: libraries/entities-renderer/src/RenderableShapeEntityItem.cpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    bool success;
+    Transform modelTransform = getTransformToCenter(success);
+    if (!success) {
+        return;
+    }
+    modelTransform.postScale(SPHERE_ENTITY_SCALE);
```
Comment (zzmp) [inline libraries/entities-renderer/src/RenderableShapeEntityItem.cpp]: Shouldn't this only occur for spheres?

File: libraries/entities-renderer/src/RenderableShapeEntityItem.h [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+#ifndef hifi_RenderableShapeEntityItem_h
+#define hifi_RenderableShapeEntityItem_h
+
+#include <ShapeEntityItem.h>
+#include <procedural/Procedural.h>
+
+#include "RenderableEntityItem.h"
+
+class RenderableShapeEntityItem : public ShapeEntityItem {
```
Comment (zzmp) [inline libraries/entities-renderer/src/RenderableShapeEntityItem.h]: The key needs to be set based on translucency so it is rendered by the correct pipeline (assuming we support translucency), or else please explicitly set an opaque key so that it is clearly rendered by a sorted pipeline.

File: libraries/entities/src/EntityItemProperties.cpp [hunk 7]
```diff
@@ -1118,7 +1121,9 @@ bool EntityItemProperties::encodeEntityEditPacket(PacketType command, EntityItem
                 APPEND_ENTITY_PROPERTY(PROP_STROKE_WIDTHS, properties.getStrokeWidths());
                 APPEND_ENTITY_PROPERTY(PROP_TEXTURES, properties.getTextures());
             }
-
+            if (properties.getType() == EntityTypes::Shape) {
+                APPEND_ENTITY_PROPERTY(PROP_TEXTURES, properties.getShape());
```
Comment (zzmp) [inline libraries/entities/src/EntityItemProperties.cpp]: Why are you using `PROP_TEXTURES`?

File: libraries/entities/src/EntityItemProperties.cpp [hunk 8]
```diff
@@ -1884,6 +1892,9 @@ QList<QString> EntityItemProperties::listChangedProperties() {
     if (queryAACubeChanged()) {
         out += "queryAACube";
     }
+    if (shapeChanged()) {
```
Comment (zzmp) [inline libraries/entities/src/EntityItemProperties.cpp]: Probably not in this PR, but this should really be a macro...

File: libraries/entities/src/EntityItemProperties.h [hunk 9]
```diff
@@ -195,6 +196,7 @@ class EntityItemProperties {
     DEFINE_PROPERTY_REF(PROP_PARENT_ID, ParentID, parentID, QUuid, UNKNOWN_ENTITY_ID);
     DEFINE_PROPERTY_REF(PROP_PARENT_JOINT_INDEX, ParentJointIndex, parentJointIndex, quint16, -1);
     DEFINE_PROPERTY_REF(PROP_QUERY_AA_CUBE, QueryAACube, queryAACube, AACube, AACube());
+    DEFINE_PROPERTY_REF(PROP_SHAPE, Shape, shape, QString, "Sphere");
```
Comment (zzmp) [inline libraries/entities/src/EntityItemProperties.h]: This is only for spheres?

File: libraries/entities/src/EntityScriptingInterface.h [hunk 10]
```diff
@@ -96,6 +96,9 @@ public slots:
     /// successful edit, if the input entityID is for an unknown model this function will have no effect
     Q_INVOKABLE QUuid editEntity(QUuid entityID, const EntityItemProperties& properties);
 
+    /// Animates a specific entitiy property
+    Q_INVOKABLE QUuid animateEntityProperty(const QUuid & entityID, const QString& propertyName, const QVariant& animationProperties);
```
Comment (zzmp) [inline libraries/entities/src/EntityScriptingInterface.h]: This animation helper should be in a separate PR.

File: libraries/entities/src/ShapeEntityItem.h [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+#ifndef hifi_ShapeEntityItem_h
+#define hifi_ShapeEntityItem_h
+
+#include "EntityItem.h"
+
+namespace entity {
+    enum Shape {
```
Comment (zzmp) [inline libraries/entities/src/ShapeEntityItem.h]: This is maintaining the same list here and in GeometryCache. Why not just alias the enum in GeometryCache?

File: scripts/developer/tests/entityEditStressTest.js [hunk 12]
```diff
@@ -15,161 +15,11 @@
...
-        createEntities();
-        Script.update.connect(updateEntities);
-        Script.scriptEnding.connect(despawnEntities);
-    }
-    Script.update.connect(init);
-})();
\ No newline at end of file
+Script.include("./entitySpawnTool.js");
+
+ENTITY_SPAWNER({
```
Comment (zzmp) [inline scripts/developer/tests/entityEditStressTest.js]: Could you add a static one, too? Like the one you used to test framerate?

File: libraries/entities/src/ShapeEntityItem.cpp [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+                                    OctreeElement::AppendState& appendState) const { 
+
+    bool successPropertyFits = true;
+    APPEND_ENTITY_PROPERTY(PROP_SHAPE, entity::stringFromShape(getShape()));
+    APPEND_ENTITY_PROPERTY(PROP_COLOR, getColor());
+    APPEND_ENTITY_PROPERTY(PROP_COLOR, getAlpha());
+}
+
+ShapeType ShapeEntityItem::getShapeType() const {
+    return SHAPE_TYPE_SPHERE;
```
Comment (zzmp) [inline libraries/entities/src/ShapeEntityItem.cpp]: :(

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp (hunk 1 above)
Comment (jherico) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: Not really.  It would be nice to minimize the amount of duplicated code between Box/Sphere and Shape entities, but we're limited in how much we can do this by the strong typing and the way the entity class hierarchy is set up.  Ideally, creating a box or a sphere would just somehow be translated into a shape entity item, but doing that in a way that doesn't confuse the communication between the entity server and the client is non-trivial.

File: libraries/entities-renderer/src/RenderableShapeEntityItem.cpp (hunk 3 above)
Comment (jherico) [inline libraries/entities-renderer/src/RenderableShapeEntityItem.cpp]: we currently don't support alpha in the procedural methods... so no.

File: libraries/entities-renderer/src/RenderableShapeEntityItem.cpp (hunk 5 above)
Comment (jherico) [inline libraries/entities-renderer/src/RenderableShapeEntityItem.cpp]: probably.

File: libraries/entities-renderer/src/RenderableShapeEntityItem.h (hunk 6 above)
Comment (jherico) [inline libraries/entities-renderer/src/RenderableShapeEntityItem.h]: The geometry cache mechanism for rendering doesn't support distinguishing between opaque and non-opaque primtives, so I'd like to leave them all opaque for now and fix the bug of not honoring the alpha in a separate PR that does more work on geometry cache.

File: libraries/entities/src/EntityItemProperties.h (hunk 9 above)
Comment (jherico) [inline libraries/entities/src/EntityItemProperties.h]: no, 'Sphere' is the default value for the shape if you don't specify one.

File: libraries/entities/src/EntityScriptingInterface.h (hunk 10 above)
Comment (jherico) [inline libraries/entities/src/EntityScriptingInterface.h]: whoops, that shouldn't be there, will remove

File: libraries/entities/src/ShapeEntityItem.h (hunk 11 above)
Comment (jherico) [inline libraries/entities/src/ShapeEntityItem.h]: the enums don't have exactly the same contents.  GeometryCache does both wire and solid drawing, so it has a line type.  ShapeEntityItem can't refer to GeometryCache because it doesn't know anything about rendering.

File: libraries/entities/src/ShapeEntityItem.cpp (hunk 13 above)
Comment (jherico) [inline libraries/entities/src/ShapeEntityItem.cpp]: as noted, this is used for physics calculations, not rendering.  I will update the code to add a comment making that clearer.

File: libraries/entities/src/EntityItemProperties.cpp (hunk 7 above)
Comment (jherico) [inline libraries/entities/src/EntityItemProperties.cpp]: fixed

File: scripts/developer/tests/entityEditStressTest.js (hunk 12 above)
Comment (jherico) [inline scripts/developer/tests/entityEditStressTest.js]: just set update interval to something really long, like 3600

**Comment by hifi-gustavo** [comment]
Build console output for commit: 358c3f7db030180f9dbdb05695182733e5537044 available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8d8f8d489cda546d74b1671f5317135d005b7e73 available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp (hunk 1 above)
Comment (ZappoMan) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: Actually... this would be trivial... all you need to do is replace the factories for "Box" and "Sphere" entity types to construct these...

Comment (ZappoMan) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: basically do this...

In `libraries/entities/src/EntityTypes.cpp`

Replace these calls:
```
REGISTER_ENTITY_TYPE(Box)
REGISTER_ENTITY_TYPE(Sphere)
```
With something like
```
    REGISTER_ENTITY_TYPE_WITH_FACTORY(Box, ShapeEntityItem::boxFactory)
    REGISTER_ENTITY_TYPE_WITH_FACTORY(Sphere, ShapeEntityItem::sphereFactory)
```

Make some static methods of `ShapeEntityItem` that construct a shape with the fixed shape type...

```
EntityItemPointer ShapeEntityItem::boxFactory(const EntityItemID& entityID, const EntityItemProperties& properties) {
    EntityItemPointer entity { new ShapeEntityItem(entityID) };
    entity->setProperties(properties);
    _shape = entity::Box; // slam the _shape to be a box no matter what the properties say
    return entity;
}


EntityItemPointer ShapeEntityItem::sphereFactory(const EntityItemID& entityID, const EntityItemProperties& properties) {
    EntityItemPointer entity { new ShapeEntityItem(entityID) };
    entity->setProperties(properties);
    _shape = entity::Sphere; // slam the _shape to be a sphere no matter what the properties say
    return entity;
}
```

Then in `libraries/entities-renderer/src/EntityTreeRenderer.cpp` change the following...

```
    REGISTER_ENTITY_TYPE_WITH_FACTORY(Box, RenderableShapeEntityItem::factory)
    REGISTER_ENTITY_TYPE_WITH_FACTORY(Sphere, RenderableSphereEntityItem::factory)
```
with
```
    REGISTER_ENTITY_TYPE_WITH_FACTORY(Box, RenderableShapeEntityItem::factory)
    REGISTER_ENTITY_TYPE_WITH_FACTORY(Sphere, RenderableShapeEntityItem::factory)
```

Those are the basic steps...

**Comment by ZappoMan** [comment]
You're going to need to make some tweaks to edit.js/entityProperties.html to support editing the properties of these shapes

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp (hunk 1 above)
Comment (ZappoMan) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: You also will probably want Sphere and Box Shapes to report their "type" as "Box" and "Sphere" instead of "Shape" to do that you need to change your `ShapeEntityItem::getProperties()` to set the "entity type" of the properties to match your geometry shape.

You can still support the idea of a Box changing it's shape... and I think that's ok... basically you could create an entity of type "Box" and then later set the Box's "shape" property to "Sphere" and then the entity would report itself as a type "Sphere"...

This seems reasonable and basically just "adds" functionality to Boxes/Spheres and doesn't regress any behavior.

**Comment by cozza13** [comment]
Interested to play with this once the performance issues are resolved, and also curious as to how much performance is improved to sphere/box primitives because of the instancing mentioned.

**Comment by hifi-gustavo** [comment]
Build console output for commit: da1506c3f7c7a3115200d8275294ba0e372066c3 available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 38128ad54cd92ac50c250a778a91891b15fafc4b available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 98fc2f9b4bf2e17c31a7ba90aadf82bbcecdb1e7 available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 99d1790f1545403b15fa31fc9f34c1607b53e515 available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 72ad612d88d978632b6de5e9f8b9ec57047f6e20 available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: cd1e910844ca3d7e9f6d0e722a6df07473fd9672 available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

File: libraries/entities/src/ShapeEntityItem.cpp [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+#include "EntityTree.h"
+#include "EntityTreeElement.h"
+#include "ShapeEntityItem.h"
+
+namespace entity {
+    static const std::vector<QString> shapeStrings { { 
+        "Triangle", 
+        "Quad", 
+        "Circle", 
+        "Cube",
```
Comment (ZappoMan) [inline libraries/entities/src/ShapeEntityItem.cpp]: Maybe this should be "Box"

File: scripts/system/html/entityProperties.html [hunk 15]
```diff
@@ -1344,6 +1362,36 @@
             <span id="property-id" class="selectable"></span>
         </div>
 
+        <div class="section-header shape-group shape-section">
+            <label>Shape</label><span>M</span>
+        </div>
+        <div class="shape-group shape-section property dropdown">
+            <label for="property-shape">Shape</label>
+            <select name="SelectShape" id="property-shape">
+                <option value="Cube">Cube</option>
```
Comment (ZappoMan) [inline scripts/system/html/entityProperties.html]: maybe it should be "box"?

File: scripts/system/html/entityProperties.html [hunk 16]
```diff
@@ -1344,6 +1362,36 @@
...
+        </div>
+        <div class="shape-group shape-section property text">
+            <label>Temp</label>
+            <input type="text" id="property-temp-1">
+        </div>
+        <div class="shape-group shape-section property text">
+            <label>Temp</label>
+            <input type="text" id="property-temp-2">
+        </div>
+<!--
```
Comment (ZappoMan) [inline scripts/system/html/entityProperties.html]: dead code?

File: scripts/system/html/entityProperties.html [hunk 17]
```diff
@@ -1344,6 +1362,36 @@
...
+        <div class="shape-group shape-section property dropdown">
+            <label for="property-shape">Shape</label>
+            <select name="SelectShape" id="property-shape">
+                <option value="Cube">Cube</option>
+                <option value="Sphere">Sphere</option>
+                <option value="Icosahedron">Icosahedron</option>
+                <option value="Tetrahedron">Tetrahedron</option>
+            </select>
+        </div>
+        <div class="shape-group shape-section property text">
```
Comment (ZappoMan) [inline scripts/system/html/entityProperties.html]: dead/debug code?

File: scripts/system/html/entityProperties.html (hunk 16 above)
Comment (jherico) [inline scripts/system/html/entityProperties.html]: while testing I discovered that if the 'combo box' expanded list overlaps a section header it appears *underneath* the section header.  So I had to make the section long enough to be able to see the combo box items.  Not sure how to deal with that as I'm not really that skilled in CSS.

File: libraries/entities/src/ShapeEntityItem.cpp (hunk 14 above)
Comment (jherico) [inline libraries/entities/src/ShapeEntityItem.cpp]: I can change the name on the entity properties to 'Box', but I'd like to keep the internal code using the actual shape names (especially since Box has connotations of a container).

File: scripts/system/html/entityProperties.html (hunk 15 above)
Comment (jherico) [inline scripts/system/html/entityProperties.html]: will fix.

File: libraries/entities/src/ShapeEntityItem.cpp (hunk 14 above)
Comment (ZappoMan) [inline libraries/entities/src/ShapeEntityItem.cpp]: But a cube is equilateral and a box implies rectangular

Comment (jherico) [inline libraries/entities/src/ShapeEntityItem.cpp]: You could make the same argument about sphere / ellipsoid.  The geometric shapes themselves are unit shapes that are subsequently scaled.

**Comment by hifi-gustavo** [comment]
Build console output for commit: c9d9024c39e5d63a6e2446772a3289fd2d464818 available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 784e9072c8f5eaabebd3f9bca507d9cbbbf6019f available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 124ce56ee0e0cb8e37acf63a5fc3fe0692fd07ec available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2c703e963cdb9780c241381948806416deffdd6a available at: https://gist.github.com/ae797f6be5f3bc19e6c9271354ea50ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7868.dmg

**Comment by Nex-Pro** [comment]
@jherico @ZappoMan Cuboid and Spheroid it is then !  :)

**Comment by Nex-Pro** [comment]
Seems stable.

@jherico In an attempt to crash this I made a few cuboids manually. Using alt+shift+drag

a few cuboids

**Comment by jherico** [comment]
## Testing 

Cubes and boxes should be create-able as before.  However, if you go into edit mode and select either a sphere or box there should now be a dropdown that allows you to chose between sphere, box, and all of the platonic solids:  tetrahedron (4 sided), octahedron (8 sided), dodecahedron (12 sided), and icosahedron (20 sided).

**Comment by kevinmthomas-carpool** [comment]
## QA Test
**QA Tester**: KevinMThomas
**QA Pass**: **QA Pass** SUCCESS
**Additional Test Results + Observations**: Tested on ```hifi://dev-demo.highfidelity.io```. Created a box entity and was able to successfully change it to a sphere, tetrahedron, octahedron, icosahedron and a dodecahedron. Changing between shapes worked and remained persistent.
